Solution Restore Service API

alpaix opened the pull request:
Preliminary version of the Restore Service API designed to accept project
restore nomination requests.

The API provides the service interface along with data structure
interfaces, needed to pass the restore data of projects references,
package references and framework assemblies.

Part of the NuGet Restore Manager design work tracked in NuGet/Home#3456.
Overall design spec is at
https://github.com/NuGet/Home/wiki/NuGet-Restore-Manager

//cc @basoundr @emgarten @joelverhagen @jainaashish @rrelyea @rohit21agrawal @yishaigalatzer

rohit21agrawal commented on src/NuGet.Clients/VisualStudio/RestoreManager/IVsSolutionRestoreService.cs:
```diff
@@ -0,0 +1,30 @@
+        /// <summary>
+        /// This property will tell if last/current restore was a success or a failure
+        /// </summary>
+        Task<bool> CurrentRestoreOperation { get; }
```
what about IsCurrentRestoreOperationSuccessful or IsCurrentRestoreOperationSuccess?

alpaix replied:
This is the CPS naming convention

joelverhagen commented on src/NuGet.Clients/VisualStudio/RestoreManager/IVsProjectReference.cs:
```diff
@@ -0,0 +1,33 @@
+        /// <summary>
+        /// Project unique name.
+        /// </summary>
+        string ProjectUniqueName { get; }
```
How will they calculate this? Can we just stick to the project path?

jainaashish commented on src/NuGet.Clients/VisualStudio/RestoreManager/IVsProjectRestoreInfo.cs:
```diff
@@ -0,0 +1,68 @@
+        /// <summary>
+        /// Framework assembly references.
+        /// </summary>
+        IReadOnlyList<IVsRestoreGraphItem> FrameworkAssemblies { get; }
```
why there are 2 separate list of IVsRestoreGraphItem? PackageReferences itself will have target framework detail, wouldn't that be sufficient?

jainaashish replied:
why there are 2 separate list of IVsRestoreGraphItem? PackageReferences itself will have target framework detail, wouldn't that be sufficient?

yishaigalatzer commented:
Where is the spec?

From: Alex Panov [mailto:notifications@github.com]
Sent: Friday, September 16, 2016 11:59 AM
To: NuGet/NuGet.Client NuGet.Client@noreply.github.com
Cc: Yishai Galatzer yigalatz@microsoft.com; Mention mention@noreply.github.com
Subject: [NuGet/NuGet.Client] Solution Restore Service API (#875)

Preliminary version of the Restore Service API designed to accept project
restore nomination requests.

The API provides the service interface along with data structure
interfaces, needed to pass the restore data of projects references,
package references and framework assemblies.

Part of the NuGet Restore Manager design work tracked in NuGet/Home#3456https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fNuGet%2fHome%2fissues%2f3456&data=02%7c01%7cyigalatz%40microsoft.com%7cdf23cc794f6145e5137508d3de638a64%7c72f988bf86f141af91ab2d7cd011db47%7c1%7c0%7c636096491654219232&sdata=JuoPGImgwMO6Ja3tdRwh6CYazHRmk0B%2bY1WCR7eqMxc%3d.
Overall design spec is at
https://github.com/NuGet/Home/wiki/NuGet-Restore-Managerhttps://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fNuGet%2fHome%2fwiki%2fNuGet-Restore-Manager&data=02%7c01%7cyigalatz%40microsoft.com%7cdf23cc794f6145e5137508d3de638a64%7c72f988bf86f141af91ab2d7cd011db47%7c1%7c0%7c636096491654219232&sdata=vmcl0Q8eg6k%2fvudFxfkaI%2bbT9psp4Pl9ldYnTgfwBTI%3d

//cc @basoundrhttps://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fbasoundr&data=02%7c01%7cyigalatz%40microsoft.com%7cdf23cc794f6145e5137508d3de638a64%7c72f988bf86f141af91ab2d7cd011db47%7c1%7c0%7c636096491654219232&sdata=uL0aHYrPuEPQNcsYGJapzhU3eZYabVZQWgRVzbPms9I%3d @emgartenhttps://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2femgarten&data=02%7c01%7cyigalatz%40microsoft.com%7cdf23cc794f6145e5137508d3de638a64%7c72f988bf86f141af91ab2d7cd011db47%7c1%7c0%7c636096491654219232&sdata=Z4i1ItjsPXwJPddEsvs2opzmXPyLwRWEIG9Qa0Mv1ug%3d @joelverhagenhttps://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fjoelverhagen&data=02%7c01%7cyigalatz%40microsoft.com%7cdf23cc794f6145e5137508d3de638a64%7c72f988bf86f141af91ab2d7cd011db47%7c1%7c0%7c636096491654219232&sdata=m8JGp21Xb3ba5kV5HQ0X1z736j3oa1tT4I7T9cQJ0vQ%3d @jainaashishhttps://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fjainaashish&data=02%7c01%7cyigalatz%40microsoft.com%7cdf23cc794f6145e5137508d3de638a64%7c72f988bf86f141af91ab2d7cd011db47%7c1%7c0%7c636096491654219232&sdata=OAscL2xPnLuTJsWMMTjBpFUU8Ya2FhHdV42A3uc2lRw%3d @rrelyeahttps://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2frrelyea&data=02%7c01%7cyigalatz%40microsoft.com%7cdf23cc794f6145e5137508d3de638a64%7c72f988bf86f141af91ab2d7cd011db47%7c1%7c0%7c636096491654219232&sdata=Ef4HwgvZ3NuyKC%2fydlZQ7eEiIZrcFJmJtssq5tYTC7I%3d @rohit21agrawalhttps://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2frohit21agrawal&data=02%7c01%7cyigalatz%40microsoft.com%7cdf23cc794f6145e5137508d3de638a64%7c72f988bf86f141af91ab2d7cd011db47%7c1%7c0%7c636096491654229240&sdata=rI2QxMIYOYO54FFJ5ZKsbLxKy20gy2ntHCXNNHiewu4%3d @yishaigalatzerhttps://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fyishaigalatzer&data=02%7c01%7cyigalatz%40microsoft.com%7cdf23cc794f6145e5137508d3de638a64%7c72f988bf86f141af91ab2d7cd011db47%7c1%7c0%7c636096491654229240&sdata=O3D5uY193G6JA3V742sfhE5GU5ZSd0o3TpI3vapffGU%3d

---

You can view, comment on, or merge this pull request online at:

  https://github.com/NuGet/NuGet.Client/pull/875https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fNuGet%2fNuGet.Client%2fpull%2f875&data=02%7c01%7cyigalatz%40microsoft.com%7cdf23cc794f6145e5137508d3de638a64%7c72f988bf86f141af91ab2d7cd011db47%7c1%7c0%7c636096491654229240&sdata=CThYpPqFpbvkFdPlTVUs2ZL5gZBxojZ0a%2bFxwLxelQs%3d

Commit Summary
-   Solution Restore Service API

File Changes
-   M src/NuGet.Clients/VisualStudio/NuGet.VisualStudio.csprojhttps://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fNuGet%2fNuGet.Client%2fpull%2f875%2ffiles%23diff-0&data=02%7c01%7cyigalatz%40microsoft.com%7cdf23cc794f6145e5137508d3de638a64%7c72f988bf86f141af91ab2d7cd011db47%7c1%7c0%7c636096491654229240&sdata=PBeq%2fKi2gvFluTdTrzCIo2BYyTdhOtsRGdAkUf4lPtc%3d (11)
-   A src/NuGet.Clients/VisualStudio/RestoreManager/IVsAssetIncludeFlags.cshttps://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fNuGet%2fNuGet.Client%2fpull%2f875%2ffiles%23diff-1&data=02%7c01%7cyigalatz%40microsoft.com%7cdf23cc794f6145e5137508d3de638a64%7c72f988bf86f141af91ab2d7cd011db47%7c1%7c0%7c636096491654239248&sdata=NQc7o1n4DGOX9Z5emSH8vlbg1gSRHnS9c%2fBxyBRkrsU%3d (44)
-   A src/NuGet.Clients/VisualStudio/RestoreManager/IVsProjectReference.cshttps://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fNuGet%2fNuGet.Client%2fpull%2f875%2ffiles%23diff-2&data=02%7c01%7cyigalatz%40microsoft.com%7cdf23cc794f6145e5137508d3de638a64%7c72f988bf86f141af91ab2d7cd011db47%7c1%7c0%7c636096491654239248&sdata=cU7cJsaVDQ9tx3Ua4UAVlzvJNjbzChmWdf2G56vyznA%3d (33)
-   A src/NuGet.Clients/VisualStudio/RestoreManager/IVsProjectRestoreInfo.cshttps://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fNuGet%2fNuGet.Client%2fpull%2f875%2ffiles%23diff-3&data=02%7c01%7cyigalatz%40microsoft.com%7cdf23cc794f6145e5137508d3de638a64%7c72f988bf86f141af91ab2d7cd011db47%7c1%7c0%7c636096491654239248&sdata=zvzYfdgGL50hxzyjzAjNLOLwGtxa5A0zL%2brpBAFllTc%3d (68)
-   A src/NuGet.Clients/VisualStudio/RestoreManager/IVsRestoreGraphItem.cshttps://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fNuGet%2fNuGet.Client%2fpull%2f875%2ffiles%23diff-4&data=02%7c01%7cyigalatz%40microsoft.com%7cdf23cc794f6145e5137508d3de638a64%7c72f988bf86f141af91ab2d7cd011db47%7c1%7c0%7c636096491654239248&sdata=z4AeLJLe0XoEeSbIR5lIEi4DM9gocBn0%2f8XOwAarsoY%3d (49)
-   A src/NuGet.Clients/VisualStudio/RestoreManager/IVsSolutionRestoreService.cshttps://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fNuGet%2fNuGet.Client%2fpull%2f875%2ffiles%23diff-5&data=02%7c01%7cyigalatz%40microsoft.com%7cdf23cc794f6145e5137508d3de638a64%7c72f988bf86f141af91ab2d7cd011db47%7c1%7c0%7c636096491654239248&sdata=D7LsVz9ZG6hlZTjz26iV%2bLdrm7BYaSzOcqiTymoLjuU%3d (30)
-   M src/NuGet.Clients/VisualStudio/project.jsonhttps://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fNuGet%2fNuGet.Client%2fpull%2f875%2ffiles%23diff-6&data=02%7c01%7cyigalatz%40microsoft.com%7cdf23cc794f6145e5137508d3de638a64%7c72f988bf86f141af91ab2d7cd011db47%7c1%7c0%7c636096491654249250&sdata=m1KV4UVoOI48MhUu5wR74U8cGOqwm9tkrJLNM113vEc%3d (3)

Patch Links:
-   https://github.com/NuGet/NuGet.Client/pull/875.patchhttps://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fNuGet%2fNuGet.Client%2fpull%2f875.patch&data=02%7c01%7cyigalatz%40microsoft.com%7cdf23cc794f6145e5137508d3de638a64%7c72f988bf86f141af91ab2d7cd011db47%7c1%7c0%7c636096491654249250&sdata=A8Mgpzkfw7mVmypi0bVp%2fAvJMoGW2SV2DNPJHRrvEys%3d
-   https://github.com/NuGet/NuGet.Client/pull/875.diffhttps://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fNuGet%2fNuGet.Client%2fpull%2f875.diff&data=02%7c01%7cyigalatz%40microsoft.com%7cdf23cc794f6145e5137508d3de638a64%7c72f988bf86f141af91ab2d7cd011db47%7c1%7c0%7c636096491654249250&sdata=VWaB4mmAxXtawA1vRhlQLu%2fkGL9zvpTOoYA9nUO08Bg%3d

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHubhttps://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fNuGet%2fNuGet.Client%2fpull%2f875&data=02%7c01%7cyigalatz%40microsoft.com%7cdf23cc794f6145e5137508d3de638a64%7c72f988bf86f141af91ab2d7cd011db47%7c1%7c0%7c636096491654249250&sdata=gJFDEeOhKH0%2fX5G7I2sqtuetJwMd6bLsQQDPxMQTBTg%3d, or mute the threadhttps://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fnotifications%2funsubscribe-auth%2fABLmt5o2bM1FA5bt05QLsHWq98tsbVpyks5qqub8gaJpZM4J_PpW&data=02%7c01%7cyigalatz%40microsoft.com%7cdf23cc794f6145e5137508d3de638a64%7c72f988bf86f141af91ab2d7cd011db47%7c1%7c0%7c636096491654249250&sdata=FfNPDqd2kprk41g7Q0vNDRM320FFFha%2bsY406ye9m0Y%3d.

jainaashish commented on src/NuGet.Clients/VisualStudio/RestoreManager/IVsRestoreGraphItem.cs:
```diff
@@ -0,0 +1,49 @@
+        /// <summary>
+        /// Version range
+        /// </summary>
+        VersionRange Version { get; } 
```
You can't use VersionRange directly in ComImport class... It will be better to string instead.

yishaigalatzer commented on src/NuGet.Clients/VisualStudio/NuGet.VisualStudio.csproj:
```diff
@@ -75,6 +75,8 @@
     <Compile Include="Extensibility\IVsFrameworkCompatibility.cs" />
```
unrelated, but please rename the project to match the generated dll name like any other project

yishaigalatzer commented on src/NuGet.Clients/VisualStudio/RestoreManager/IVsAssetIncludeFlags.cs:
```diff
@@ -0,0 +1,44 @@
+namespace NuGet.VisualStudio
+{
+    /// <summary>
+    /// Asset include type flags
```
Make the comments more meaningful and useful to readers of this code. I'm also not sure we should actually have flags as enums, because of the potential for change here (adding flags, changing behaviors)

alpaix replied:
Agree. Eventually I will. Just want to share interfaces as quickly as possible to collect some feedback.

yishaigalatzer commented on src/NuGet.Clients/VisualStudio/RestoreManager/IVsProjectReference.cs:
```diff
@@ -0,0 +1,33 @@
+    /// </summary>
+    [ComImport]
+    [Guid("91db043c-6766-4d2c-9885-6861457bac76")]
+    public interface IVsProjectReference
```
This is locking the data model down pretty hard, do we need to create a new interface every time a new piece of data needs to be copied over, while maintaining this old one?

yishaigalatzer commented on src/NuGet.Clients/VisualStudio/RestoreManager/IVsProjectRestoreInfo.cs:
```diff
@@ -0,0 +1,68 @@
+        /// <summary>
+        /// User packages folder path.
+        /// </summary>
+        string PackagesPath { get; }
```
Need to be more specific, I can't tell if this is the packages.config folder or the global one scoped.

What is the purpose of this interface? Producer/consumer?

yishaigalatzer commented:
I'd like to discuss the layering before accepting this PR, it is not reflected in the spec, and this PR seems to create a very rigid interface with "something" that is not specified here, so I'm just guessing it is with CPS, which blocks adding extra metadata without extra handshakes

Commit:
Avoid adding nuget actions for dependencies if dependency behavior is set to ignore.

Fix https://github.com/NuGet/Home/issues/1298

alpaix committed:
Solution Restore Service API

3rd clean version of the Restore Service API designed to accept project
restore nomination requests.

The API provides the service interface along with data structure
interfaces, needed to pass the restore data of projects references,
package references and framework assemblies.

Part of the NuGet Restore Manager design work tracked in NuGet/Home#3456.
Overall design spec is at
https://github.com/NuGet/Home/wiki/NuGet-Restore-Manager

alpaix commented:
Incorporated notes from the feedback. Now the API is based on properties rather than strongly typed classes. Also isolated the API in a separate project.

Contains some noise caused by moving `NuGet.VisualStudio` project as requested.

//cc @joelverhagen @jainaashish @rohit21agrawal @rrelyea @basoundr @yishaigalatzer

basoundr commented:
`IntermediateOutputPath`  or `OutputPath` is missing. Apart  from that, rest looks good to me

basoundr commented:
Adding @Srivatsn @davkean

Commit:
Adding unit test coverage to locals command

mishra14 committed:
Adding conditional references to ComponentModelHost.dll (#884)

Adding conditional references to ComponentModelHost.dll for 7 projects in the NuGetClients solution

Commit:
nuget telemetry events for VS15

joelverhagen committed:
Switch dev over to 3.6.0-rc, since we are about spin off a 3.6.0-beta2 branch (from dev-msbuild) (#887)

joelverhagen committed:
Bump NuGet.Core version to 2.14.0-rtm-832 to have a latest bug fix (#886)

NuGet.Core bug fix was https://github.com/NuGet/Home/issues/3410

alpaix commented:
Closing the PR in favor of a new clean revision of API #888.
